src/libraries/CalldataDecoder.sol

```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.23;
+
+import {Call} from "./CallLib.sol";
+
+library CalldataDecoder {
+    using CalldataDecoder for bytes;
+    /// @notice equivalent to SliceOutOfBounds.selector, stored in least-significant bits
+
+    error SliceOutOfBounds();
+
+    /// @notice mask used for offsets and lengths to ensure no overflow
+    /// @dev no sane abi encoding will pass in an offset or length greater than type(uint32).max
+    ///      (note that this does deviate from standard solidity behavior and offsets/lengths will
+    ///      be interpreted as mod type(uint32).max which will only impact malicious/buggy callers)
+    uint256 constant OFFSET_OR_LENGTH_MASK = 0xffffffff;
+    uint256 constant OFFSET_OR_LENGTH_MASK_AND_WORD_ALIGN = 0xffffffe0;
+
+    uint256 constant SLICE_ERROR_SELECTOR = 0x3b99b53d;
+
+    /// @notice equivalent to abi.decode(data, (uint256, bytes)) in calldata
+    /// @return nonce The nonce value.
+    /// @return signature The signature to verify.
+    function decodeOpdata(bytes calldata data) internal pure returns (uint256 nonce, bytes calldata signature) {
+        assembly {
+            nonce := calldataload(data.offset)
+        }
+        signature = data.toBytes(1);
+    }
```

Comment:
I don't really like having biz logic here in the libraries, seeing as we may want to handle decoding depending on the mode, maybe we should make this more generic for types

Reply:
you mean like call it decodeUint256Bytes or something?

Reply:
this might be nicer if implemented in the ModeDecoder, since its possible that more than one mode will have the same opData decoding.

`function decodeOpdata(bytes32 mode, bytes calldata data)`

and then do conditional branching within the function, you have access to `isBatchedCall`, etc. too